community/kernel/src/test/java/org/neo4j/graphdb/GraphDatabaseServiceTest.java

```diff
@@ -71,44 +75,45 @@ public void givenDatabaseAndStartedTxWhenShutdownThenWaitForTxToFinish() throws
         final GraphDatabaseService db = new TestGraphDatabaseFactory().newImpermanentDatabase();
 
         // When
-        final CountDownLatch started = new CountDownLatch( 1 );
-        Executors.newSingleThreadExecutor().submit( new Runnable()
+        Barrier.Control barrier = new Barrier.Control();
+        Future<Object> txFuture = t2.execute( state ->
         {
-            @Override
-            public void run()
+            try ( Transaction tx = db.beginTx() )
             {
-                try ( Transaction tx = db.beginTx() )
-                {
-                    started.countDown();
-
-                    try
-                    {
-                        Thread.sleep( 2000 );
-                    }
-                    catch ( InterruptedException e )
-                    {
-                        throw new RuntimeException( e );
-                    }
-
-                    db.createNode();
-                    tx.success();
-                }
-                catch ( Throwable e )
-                {
-                    e.printStackTrace();
-                }
+                barrier.reached();
+                db.createNode();
+                tx.success();
             }
+            return null;
         } );
 
-        started.await();
-        db.shutdown();
+        // i.e. wait for transaction to start
+        barrier.await();
+
+        // now there's a transaction open, blocked on continueTxSignal
+        Future<Object> shutdownFuture = t3.execute( state ->
+        {
+            db.shutdown();
+            return null;
+        } );
+        t3.get().waitUntilWaiting( location -> location.isAt( DatabaseAvailability.class, "shutdown" ) );
```

Comment:
Yes exactly, this technique has existed for years in the neo4j test code base, although many tests don't take advantage of it, instead waiting an arbitrary amount of seconds, catch a TimeoutException and interpret that is the same result. This technique is actually better in every way because we can very specifically determine where the wait happens and it can be detected much faster.